### ISSUE TITLE
randtest and fourthcorner statistics for partial RLQ? Error in UseMethod("randtest" / "fourthcorner.rlq") : no applicable method for 'randtest' applied to an object of class "c('within', 'dudi')"

### ISSUE THREAD

**Issue by maurinka** [adeverse/ade4#45]
How can I evaluate global significance of a partial RLQ (wca.rlq), and calculate the contributions of the individual variables to the RLQ axes just as described by Wesuls et al. (2012, doi: 10.1111/j.1654-1103.2011.01342.x)?

i.e. what would be the corresponding functions for:

> nrepet = 999
> testrlq.env <- randtest(rlq.env, modeltype = 6, nrepet = nrepet)

> randtest(wrlq_crop, modeltype = 6, nrepet = nrepet) # doesnt work
Error in UseMethod("randtest") : 
  no applicable method for 'randtest' applied to an object of class "c('within', 'dudi')"

and 

> testQaxes.comb.env <- fourthcorner.rlq(rlq.env, modeltype = 6, typetest = "Q.axes", nrepet = nrepet, p.adjust.method.G = "fdr", p.adjust.method.D = "fdr")
> testRaxes.comb.env <- fourthcorner.rlq(rlq.env, modeltype = 6, typetest = "R.axes", nrepet = nrepet, p.adjust.method.G = "fdr",  p.adjust.method.D = "fdr")

> fourthcorner.rlq(wrlq_crop, modeltype = 6,  typetest = "R.axes", nrepet = nrepet, p.adjust.method.G = "fdr", p.adjust.method.D = "fdr")
Error in fourthcorner.rlq(wrlq_crop, modeltype = 6, typetest = "R.axes",  : 
  Object of class 'rlq' expected

**Comment by sdray** [comment]
Sorry for not replying before but was very busy these last days. The method is not properly implemented in ade4 and you should use a small trick. I give you some code from the original paper showing how to run the test:

```
## we assume that species abundances are stored in lnspec, traits in trait. Run  standard rlq and results stored in rlqTR
## habitat[,1] is a factor with a covariable that we want to remove the effect with withn-class rlq

wit1 <- within.rlq(rlqTR,habitat[,1], scannf = TRUE, nf = 2)

#then fourth-corner on the site scores of within RLQ
four1<-fourthcorner(wit1$lR,lnspec,trait,modeltype=2, nrepet=99) ## the trick : use output of within-RLQ wit1$lR instead of original environmental variables

summary(four1)
plot(four1,type = "G")
plot(four1,type = "D")

four2<-fourthcorner(wit1$lR,lnspec,trait,modeltype=4, nrepet=99)
summary(four2)
plot(four2,type = "G")
plot(four2,type = "D")

four3 <- combine.4thcorner(four1,four2)
summary(four3)
plot(four3,alpha=sqrt(0.05),type = "G")
plot(four3,alpha=sqrt(0.05),type = "D")

print(four3)
```

**Comment by maurinka** [comment]
thank you so much.


what about the functions randtest() and fourthcorner.rlq() ? i tried the same trick, but it still doesn't work. also, do i need to include the FDR in each model before the combination or only when combining them?

**Comment by sdray** [comment]
yep. so for rlq, the idea is to apply pca (without centering/scaling) onoutputs of within class analysis performed on environmental variables:

```
pca_env <- dudi.pca(env_table)
wit_env <- wca(pca_env, habitat[,1])
pca_wit <- dudi.pca(wit$tab, scale = FALSE, center = FALSE, row. w= ca_abund$lw)
rlq(pca_wit, ca_abund, pca_trait)


```

Cheers

**Comment by maurinka** [comment]
sorry, i dont get it. id would be amazing (and probably more efficient) of we could quickly talk. can I send you my phonenumber by email?
